internal/multiparty/manager.go

```diff
@@ -0,0 +1,167 @@
+	TerminateContract(ctx context.Context, contracts *core.FireFlyContracts, termination *blockchain.Event) (err error)
+
+	// GetNetworkVersion returns the network version of the active FireFly contract
+	GetNetworkVersion() int
+
+	// SubmitBatchPin sequences a batch of message globally to all viewers of a given ledger
+	SubmitBatchPin(ctx context.Context, nsOpID string, signingKey string, batch *blockchain.BatchPin) error
+
+	// SubmitNetworkAction writes a special "BatchPin" event which signals the plugin to take an action
+	SubmitNetworkAction(ctx context.Context, nsOpID string, signingKey string, action core.NetworkActionType) error
+
+	core.Named
+}
+
+type Contract struct {
+	Location   *fftypes.JSONAny
+	FirstEvent string
+}
+
+type multipartyManager struct {
+	ctx            context.Context
+	blockchain     blockchain.Plugin
+	contracts      []Contract
+	namespace      string
+	activeContract struct {
+		location       *fftypes.JSONAny
+		mux            sync.Mutex
+		firstEvent     string
+		networkVersion int
+		subscription   string
+	}
+}
+
+func NewMultipartyManager(ctx context.Context, ns string, contracts []Contract, bi blockchain.Plugin) Manager {
+	return &multipartyManager{
+		ctx:        ctx,
+		namespace:  ns,
+		contracts:  contracts,
+		blockchain: bi,
+	}
+}
+
+func (mm *multipartyManager) ConfigureContract(ctx context.Context, contracts *core.FireFlyContracts) (err error) {
+	log.L(ctx).Infof("Resolving FireFly contract at index %d", contracts.Active.Index)
+	location, firstEvent, err := mm.resolveFireFlyContract(ctx, contracts.Active.Index)
+	if err != nil {
+		return err
+	}
+	version, err := mm.blockchain.GetNetworkVersion(ctx, location)
+	if err != nil {
+		return err
+	}
+
+	subID, err := mm.blockchain.AddFireflySubscription(ctx, mm.namespace, location, firstEvent)
+	if err == nil {
+		mm.activeContract.mux.Lock()
+		mm.activeContract.location = location
+		mm.activeContract.firstEvent = firstEvent
+		mm.activeContract.networkVersion = version
+		mm.activeContract.subscription = subID
+		mm.activeContract.mux.Unlock()
+		contracts.Active.Info = fftypes.JSONObject{
+			"location":     location,
+			"firstEvent":   firstEvent,
+			"subscription": subID,
+		}
+	}
+	return err
+}
+
+func (mm *multipartyManager) resolveFireFlyContract(ctx context.Context, contractIndex int) (location *fftypes.JSONAny, firstEvent string, err error) {
+	if len(mm.contracts) > 0 {
+		if contractIndex >= len(mm.contracts) {
+			return nil, "", i18n.NewError(ctx, coremsgs.MsgInvalidFireFlyContractIndex, fmt.Sprintf("%s.multiparty.contracts[%d]", mm.namespace, contractIndex))
+		}
+		active := mm.contracts[contractIndex]
+		location = active.Location
+		firstEvent = active.FirstEvent
+
+		switch firstEvent {
+		case string(core.SubOptsFirstEventOldest):
+			firstEvent = "0"
+		case string(core.SubOptsFirstEventNewest):
+			firstEvent = "latest"
```

Comment:
This looks blockchain-specific. Feels like `firstEvent` should be passed on down and let the blockchain determine how to spell the special cases for `SubOptsFirstEventOldest` and `SubOptsFirstEventNewest`.